### PR TITLE
FINERACT-1968: Advanced payment allocation - Default allocation

### PR DESCRIPTION
**Pull request by adamsaghy**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java [hunk 1]
```diff
@@ -74,20 +92,285 @@ protected Money handleRefundTransactionPaymentOfInstallment(LoanRepaymentSchedul
...
+            }
+        }
+
+        for (final LoanRepaymentScheduleInstallment currentInstallment : installments) {
+            currentInstallment.resetDerivedComponents();
+            currentInstallment.updateDerivedFields(currency, disbursementDate);
+        }
+
+        // TODO: Remove this reprocess and add the charges to the installment in chronological order
+        final LoanRepaymentScheduleProcessingWrapper wrapper = new LoanRepaymentScheduleProcessingWrapper();
```
Comment (ruchiD) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java:114]: Hi Adam, one question here, as charges are added to installment according to charge due date, what changes we are proposing here for following the chronological order.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java [hunk 2]
```diff
@@ -74,20 +92,285 @@ protected Money handleRefundTransactionPaymentOfInstallment(LoanRepaymentSchedul
...
+        List<LoanPaymentAllocationRule> paymentAllocationRules = loanTransaction.getLoan().getPaymentAllocationRules();
+        LoanPaymentAllocationRule defaultPaymentAllocationRule = paymentAllocationRules.stream()
+                .filter(e -> PaymentAllocationTransactionType.DEFAULT.equals(e.getTransactionType())).findFirst().orElseThrow();
+        LoanPaymentAllocationRule paymentAllocationRule = paymentAllocationRules.stream()
+                .filter(e -> loanTransaction.getTypeOf().equals(e.getTransactionType().getLoanTransactionType())).findFirst()
+                .orElse(defaultPaymentAllocationRule);
+        Balances balances = new Balances(zero, zero, zero, zero);
+        for (PaymentAllocationType paymentAllocationType : paymentAllocationRule.getAllocationTypes()) {
+            FutureInstallmentAllocationRule futureInstallmentAllocationRule = paymentAllocationRule.getFutureInstallmentAllocationRule();
+            LoanRepaymentScheduleInstallment currentInstallment = null;
```
Comment (ruchiD) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java:190]: What is default future installment allocation rule if not defined for a transaction type.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java (hunk 1 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java:114]: The legacy behaviour was to add all the available charges to the repayment schedules based on the due date. The problem with this, if the charge was added later than a repayment, during the reprocessing the repayment might was paying off that charge. Chronologically when we are processing the repayment and allocating it we should not consider the charges that were added just later. 

In TLDR: when we are reprocessing the transactions, we should reprocessing the charges as well and add them to the installments one by one based on whether it was added before or after than the processed transaction.

I hope it helps to understand. A follow up story should cover this capability.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java (hunk 2 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java:190]: It must be defined. If it was not set, there shall be a validation error when this allocation rules were set to the loan product.

**Review by ruchiD — COMMENTED** [review]
LGTM

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java (hunk 1 above)
Comment (ruchiD) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java:114]: Okay.

**Comment by jdailey** [comment]
This needs a section in the documentation

**Pull request merged**